src/wobjectcpp.h

```diff
@@ -0,0 +1,124 @@
+#if __cplusplus > 201700L
+        -> MetaMethodInfoBuilder<F, (fs | ... | Flags), IC, ParamTypes, ParamNames> {
+#else
+        -> MetaMethodInfoBuilder<F, w_internal::summed<fs ..., Flags>, IC, ParamTypes, ParamNames> {
+#endif
+        return {name, func, paramTypes, paramNames};
+    }
+    template<class IC2>
+    constexpr auto setIntegralConstant() const -> MetaMethodInfoBuilder<F, Flags, IC2, ParamTypes, ParamNames> {
+        return {name, func, paramTypes, paramNames};
+    }
+
+    constexpr auto build() const -> w_internal::MetaMethodInfo<F, Flags, IC, ParamTypes, ParamNames> {
+        return {func, name, paramTypes, paramNames};
+    }
+};
+
+template<typename F>
+constexpr auto makeSignalBuilder(StringView name, F func) {
+    return MetaMethodInfoBuilder<F, W_MethodType::Signal.value>{name, func};
+}
+
+
+} // namespace w_cpp
+
+/// \macro W_CPP_PROPERTY(callback)
+/// allows to create multiple properties from a templated structure using regular C++.
+///
+/// example usage:
+///     template<size_t I>
+///     struct MyProperties;
+///     template<>
+///     struct MyProperties<0> {
+///         constexpr static auto property = w_cpp::makeProperty<QString>(w_cpp::viewLiteral("name"), w_cpp::viewLiteral("QString"))
+///             .setGetter(&Example::getName);
+///     };
+///     W_CPP_PROPERTY(MyProperties)
+///
+/// NOTE: you have to ensure that the struct is only valid for some `I`s.
+#define W_CPP_PROPERTY(a) \
+    static constexpr size_t W_MACRO_CONCAT(a,_O) = w_internal::stateCount<__COUNTER__, w_internal::PropertyStateTag, W_ThisType**>; \
+    template<size_t I> \
+    friend constexpr auto w_state(w_internal::Index<I>, w_internal::PropertyStateTag, W_ThisType**) W_RETURN((a<I-W_MACRO_CONCAT(a,_O)>::property))
+
+/// \macro W_CPP_SIGNAL(callback)
+/// allows to register multiple signals from a templated structure using regular C++.
+///
+/// example usage:
+///     template<size_t I>
+///     struct MySignals;
+///     template<>
+///     struct MySignals<0> {
+///         constexpr static auto signal = w_cpp::makeSignalBuilder(w_cpp::viewLiteral("nameChanged"), &Example::nameChanged).build();
+///     };
+///     W_CPP_SIGNAL(MySignals)
+///
+/// NOTE: you have to ensure that the struct is only valid for some `I`s.
+#define W_CPP_SIGNAL(a) \
+    static constexpr size_t W_MACRO_CONCAT(a,_O) = w_internal::stateCount<__COUNTER__, w_internal::SignalStateTag, W_ThisType**>; \
+    template<size_t I> \
+    friend constexpr auto w_state(w_internal::Index<I>, w_internal::SignalStateTag, W_ThisType**) W_RETURN((a<I-W_MACRO_CONCAT(a,_O)>::signal))
+
+/// \macro W_CPP_SIGNAL_IMPL(type, callback, index)
+/// allows to implement a signal
+///
+/// example usage:
+///     template<size_t I>
+///     void notifyPropertyChanged() W_CPP_SIGNAL_IMPL(decltype (&tst_CppApi::notifyPropertyChanged<I>), MySignals, I, 0)
+///
+#define W_CPP_SIGNAL_IMPL(type, a, i, ...) { \

```

Comment:
I am not really understanding the use case.

This does not really works if the signal has different type.

Perhaps the best way to emit the signal is to simply call QMetaObject::activate somehow

Reply:
The `W_SIGNAL` macro generates both an implementation of the signal function and the registration.

The registration is easily done with the constexpr hook. But the implementation part would be missing.
This macro allows the generation of the signal function body. Inspired by [wobjectdefs.h#L788](https://github.com/woboq/verdigris/blob/master/src/wobjectdefs.h#L788).

It's possible that there is a bug I don't see. But it works well for our use case of property changed signals without any arguments.